src/hardware/dac.rs

```diff
@@ -0,0 +1,156 @@
+///! Thermostat DAC driver
+///!
+///! This file contains the driver for the 4 Thermostat DAC output channels.
+///! To convert a 20 bit word into an analog current Thermostat uses a DAC to
+///! convert the word into a voltage and a subsequent TEC driver IC that produces
+///! a current proportional to the DAC voltage.
+///!
+///! The 4 channel DAC ICs share an SPI bus and are addressed using individual "sync"
+///! signals, similar to a chip select signal.
+///! The TEC driver ICs feature a shutdown mode controlled by a shutdown signal and
+///! current limits controlled by another input voltage. The shutdown signal is aggregated
+///! into the DAC driver using a gpio, while the current limits make use of the Thermostat
+///! PWM driver.
+///! DAC datasheet: https://www.analog.com/media/en/technical-documentation/data-sheets/AD5680.pdf
+///! TEC driver datasheet: https://datasheets.maximintegrated.com/en/ds/MAX1968-MAX1969.pdf
+///!
+use super::{
+    hal::{
+        gpio::{gpioc::*, gpiog::*, Alternate, Output, PushPull, AF6},
+        hal::{blocking::spi::Write, digital::v2::OutputPin},
+        prelude::*,
+        rcc::{rec, CoreClocks},
+        spi::{Enabled, NoMiso, Spi, MODE_1},
+        stm32::SPI3,
+        time::MegaHertz,
+    },
+    MAXCODE, R_SENSE, VREF_DAC, VREF_TEC,
+};
+
+use super::Channel;
+
+// Note: 30MHz clock valid according to DAC datasheet. This lead to spurious RxFIFO overruns on the STM side when probing the spi clock with a scope probe.
+const SPI_CLOCK: MegaHertz = MegaHertz(8);
+
+/// Convert TEC drive current to dac code.
+fn i_to_dac(i: f32) -> u32 {
+    let v = (i * 10.0 * R_SENSE) + VREF_TEC;
```

Comment:
I would still try to keep the DAC mod self-contained as much as possible. It seems sensible to let the API here convert from SI volts (at the DAC output) to DAC code. I.e. do the conversion from TEC current to PWM driver input voltage elsewhere.
Then there should be no need for shared constants.

Reply:
stays here for now